lib/ace/ext/papeeria/spellchecker.coffee

```diff
@@ -1,39 +1,119 @@
+        showPopup = (options) =>
+            AcePopup = require("ace/autocomplete/popup").AcePopup
+            util = require("ace/autocomplete/util")
+            HashHandler = require("ace/keyboard/hash_handler").HashHandler
+            event = require("ace/lib/event")
+
+            @popup = new AcePopup(document.body ? document.documentElement)
+            @popup.setData(options)
+            @popup.setTheme(editor.getTheme())
+            @popup.setFontSize(editor.getFontSize())
+
+            @commands = {
+                # TODO
+                # Doesn't work yet for some reason
+                "Esc": (editor) ->
+                    @popup.hide()
+                    editor.keyBinding.removeKeyboardHandler(@keyboard);
+                    console.log("esc")
+
+            }
+            @keyboard = new HashHandler()
+            @keyboard.bindKeys(@commands)
+            editor.keyBinding.addKeyboardHandler(@keyboard)
+
+            # This part is like "magic, do not touch" thing. Was copied from autocomplete.js,
+            # seems to set things up in a certain way so popup window opens under the cursor.
+            renderer = editor.renderer
+            lineHeight = renderer.layerConfig.lineHeight
+            pos = renderer.$cursorLayer.getPixelPosition(this.base, true)
+            pos.left -= @popup.getTextLeftOffset()
+            rect = editor.container.getBoundingClientRect()
+            pos.top += rect.top - renderer.layerConfig.offset
+            pos.left += rect.left - editor.renderer.scrollLeft
+            pos.left += renderer.gutterWidth
+
+            @popup.show(pos, lineHeight)
+            return
+
+        # Convert an array of string to popup eligible structure.
+        # @param {Array} corrections: array of strings with substitution options.
+        # @returns {Array}: array of jsons, actually.
+        convertCorrectionList = (corrections) ->
+            options = []
+            options.push({caption: item, value: item}) for item in corrections
+            return options
+
+        # Get the word under the cursor.
+        # @returns {String}
+        extractWord = ->
+            session = editor.session
+            row = editor.getCursorPosition().row
+            col = editor.getCursorPosition().column
+            wordPosition = session.getAWordRange(row, col)
+            start = wordPosition.start.column
+            end = wordPosition.end.column
+            # getAWordRange returns start and end positions with a trailing
+            # whitespace at the end (if there's a one), that's why replace is used
+            return session.getLine(row).substring(start, end).replace(/\s\s*$/, '')
+
+        # Check if current word is in corrections list.
+        # Call showPopup if it's present, do nothing otherwise.
+        tryPopup = ->
+            word = extractWord()
+            spellChecker = new SpellChecker()
+            correctionsList = spellChecker.getJson()
```

Comment:
perfect decapsulation :) you may want to remove method `check` if you're using spellchecker internals anyway